snakemake/executors/__init__.py

```diff
@@ -976,8 +980,38 @@ def __init__(
             )
 
     def cancel(self):
-        logger.info("Will exit after finishing currently running jobs.")
-        self.shutdown()
+        # maximum number of jobs to cancel at once
+        max_mcancel = 1000
+
+        def _chunks(lst, n):
+            """Yield successive n-sized chunks from lst."""
+            for i in range(0, len(lst), n):
+                yield lst[i : i + n]
+
+        if self.cancelcmd or self.mcancelcmd:  # We have --cluster-[m]cancel
+            # Enumerate job IDs and create chunks.  Set size to 1 if not mcancel, else
+            # limit to a reasonable size (few cancel commands but not too long command
+            # line).
+            jobids = [j.jobid for j in self.active_jobs]
+            if self.mcancelcmd:
+                cmd = self.mcancelcmd
+                chunks = list(_chunks(jobids, max_mcancel))
+            else:
+                cmd = self.cancelcmd
+                chunks = list(_chunks(jobids, 1))
+            # Go through the chunks and cancel the jobs.
+            for chunk in chunks:
+                try:
+                    subprocess.Popen([cmd] + chunk, shell=False)
+                except:
+                    # This is common - logging a warning would probably confuse the user.
```

Comment:
True, but we should collect the number of failed cancel commands, and issue a warning at the end that tells the user how many of them failed (if that number is greater than 0).

Reply:
Agreed, this was here before but will be fixed.